## Mini Project

In [ ]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Setting Working Directory

In [ ]:
os.chdir("D:/Mini Project/6. Classification")

### Loading the Dataset


In [ ]:
bank = pd.read_csv("PL_XSELL.csv")

In [ ]:
bank.info()

### Removig columng not required for making the model

In [ ]:
bank_final=bank.drop(columns=['AGE_BKT','ACC_OP_DATE','NO_OF_L_CR_TXNS','NO_OF_L_DR_TXNS','TOT_NO_OF_L_TXNS',
                              'AMT_OTH_BK_ATM_USG_CHGS','AMT_MIN_BAL_NMC_CHGS','NO_OF_IW_CHQ_BNC_TXNS',
                              'NO_OF_OW_CHQ_BNC_TXNS','AVG_AMT_PER_ATM_TXN','AVG_AMT_PER_CSH_WDL_TXN','AVG_AMT_PER_CHQ_TXN',
                              'AVG_AMT_PER_NET_TXN','AVG_AMT_PER_MOB_TXN','random'])

In [ ]:
bank_final.info()

### Split datasets into train and test datasets

In [ ]:
from sklearn.cross_validation import train_test_split

In [ ]:
train,test = train_test_split( bank_final, test_size = 0.3)

In [ ]:
train.count()

In [ ]:
test.count()

### Splitting into features and response variables

In [ ]:
X_train =  train[['AGE', 'GENDER', 'BALANCE','OCCUPATION','SCR', 'HOLDING_PERIOD', 'ACC_TYPE', 'LEN_OF_RLTN_IN_MNTH',
                 'NO_OF_BR_CSH_WDL_DR_TXNS','NO_OF_ATM_DR_TXNS',
                 'NO_OF_NET_DR_TXNS','NO_OF_MOB_DR_TXNS','NO_OF_CHQ_DR_TXNS','FLG_HAS_CC','AMT_ATM_DR','AMT_BR_CSH_WDL_DR',
                 'AMT_CHQ_DR','AMT_NET_DR','AMT_MOB_DR','AMT_L_DR','FLG_HAS_ANY_CHGS','FLG_HAS_NOMINEE','FLG_HAS_OLD_LOAN']]

In [ ]:
y_train = train["TARGET"]

Categorical Variable to Numerical Variables

In [ ]:
X_train = pd.get_dummies(X_train)
X_train.columns

In [ ]:
X_train.head()

In [ ]:
X_test=test[['AGE', 'GENDER', 'BALANCE','OCCUPATION','SCR', 'HOLDING_PERIOD', 'ACC_TYPE', 'LEN_OF_RLTN_IN_MNTH',
                 'NO_OF_BR_CSH_WDL_DR_TXNS','NO_OF_ATM_DR_TXNS',
                 'NO_OF_NET_DR_TXNS','NO_OF_MOB_DR_TXNS','NO_OF_CHQ_DR_TXNS','FLG_HAS_CC','AMT_ATM_DR','AMT_BR_CSH_WDL_DR',
                 'AMT_CHQ_DR','AMT_NET_DR','AMT_MOB_DR','AMT_L_DR','FLG_HAS_ANY_CHGS','FLG_HAS_NOMINEE','FLG_HAS_OLD_LOAN']]
y_test= test["TARGET"]

In [ ]:
X_test= pd.get_dummies(X_test)

In [ ]:
print(type(X_train) , type(y_train))

In [ ]:
X_train.count()

In [ ]:
X_test.count()

In [ ]:
y_train.value_counts()

In [ ]:
Response_percentage = 1801/14000 * 100
Response_percentage


Only 12.86 % are 1’s here. So the data is unbalanced, need to oversample the data to make it balanced.

### Decision Tree

In [254]:
from sklearn.tree import DecisionTreeClassifier

Setting the parameter

In [255]:
clf = DecisionTreeClassifier(criterion = "gini" , 
                             min_samples_split = 100,
                             min_samples_leaf = 10,
                             max_depth = 50)

Calling the fit function to build the tree

In [256]:
clf.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=50,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=10, min_samples_split=100,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [304]:
import pydotplus
import graphviz
import io
from sklearn.tree import export_graphviz
from scipy import misc
from matplotlib import pyplot as plt
f = io.StringIO()

In [308]:
dot_data=tree.export_graphviz(clf,out_file=None)
graph = graphviz.Source(dot_data)
graph.render("Cart")

Checking the nodes

In [309]:
Nodes = pd.DataFrame(clf.tree_.__getstate__()["nodes"])
Nodes.head()

left_child  right_child  feature  threshold  impurity  n_node_samples  \
0           1          336        3       18.5  0.224400           14000   
1           2          285       25        0.5  0.267902            8736   
2           3          124       11     7650.0  0.242510            7196   
3           4           83       10        0.5  0.174483            3262   
4           5           78       16  1528705.0  0.142461            2319   

   weighted_n_node_samples  
0                  14000.0  
1                   8736.0  
2                   7196.0  
3                   3262.0  
4                   2319.0

### Finding the variable with more importance

In [310]:
feature_importance = pd.DataFrame([X_train.columns,
                               clf.tree_.compute_feature_importances()])
feature_importance.T 

0           1
0                        AGE   0.0441409
1                    BALANCE   0.0968526
2                        SCR    0.130252
3             HOLDING_PERIOD    0.113209
4        LEN_OF_RLTN_IN_MNTH   0.0509812
5   NO_OF_BR_CSH_WDL_DR_TXNS   0.0239865
6          NO_OF_ATM_DR_TXNS   0.0220064
7          NO_OF_NET_DR_TXNS   0.0144929
8          NO_OF_MOB_DR_TXNS  0.00159868
9          NO_OF_CHQ_DR_TXNS   0.0187131
10                FLG_HAS_CC   0.0391015
11                AMT_ATM_DR   0.0648283
12         AMT_BR_CSH_WDL_DR   0.0707139
13                AMT_CHQ_DR   0.0404548
14                AMT_NET_DR    0.046362
15                AMT_MOB_DR   0.0554876
16                  AMT_L_DR   0.0686454
17          FLG_HAS_ANY_CHGS  0.00622358
18           FLG_HAS_NOMINEE  0.00112519
19          FLG_HAS_OLD_LOAN           0
20                  GENDER_F           0
21                  GENDER_M  0.00507665
22                  GENDER_O  0.00861045
23           OCCUPATION_PROF  0.00180747
24            OCCUPATION_SAL  0.00650224
25       OCCUPATION_SELF-EMP   0.0462418
26           OCCUPATION_SENP  0.00843603
27               ACC_TYPE_CA  0.00866891
28               ACC_TYPE_SA   0.0054811

Here , SCR = 0.1302, which is the most important variable in the dataset

### Checking how good is the model,

Predicting Y values with training data set

In [311]:
pred_y_train = clf.predict(X_train )
pred_y_train

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

Let us see the classification accuracy of our model

In [312]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_curve
from sklearn.metrics import auc

In [313]:
score = accuracy_score(y_train, pred_y_train)
score

0.8871428571428571

Here the got accuracy is 88.71%, which is same as that without model. So here classification accuracy is not a good measure.

In [314]:
y_train_prob = clf.predict_proba(X_train)
y_train_prob

array([[1.        , 0.        ],
       [0.98734177, 0.01265823],
       [0.87323944, 0.12676056],
       ...,
       [1.        , 0.        ],
       [0.97916667, 0.02083333],
       [1.        , 0.        ]])

#### Checking for AUC,

In [315]:
fpr, tpr, thresholds = roc_curve(y_train, y_train_prob[:,1])
auc(fpr, tpr)


0.8936883636725268

Here AUC is around 0.89, so the model is good

### Checking accuracy with test data set

In [316]:
pred_y_test = clf.predict(X_test)
pred_y_test

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [317]:
score_h = accuracy_score(y_test, pred_y_test)
score_h

0.8726666666666667

Here accuracy is 87.26%, the model accuracy is not much deviating from that of training data. So the model is not overfitting


### AUC for test data

In [318]:
y_test_prob = clf.predict_proba(X_test)
fpr, tpr, thresholds = roc_curve(y_test, y_test_prob[:,1])
auc(fpr, tpr)

0.8136965957840429

AUC has reduced in the test data.

In [319]:
(0.8936-0.8136)/0.8936

0.08952551477170989

That is 8.95 % variation in AUC compared to that of training data, that is the model is slightly over fitting.

### Now doing cross validation to overcome over fitting

In [320]:
from sklearn.cross_validation import cross_val_score
scores = cross_val_score(clf, X_train , y_train, cv = 10, scoring='roc_auc')

scores.mean()


0.7825178682015925

In [321]:
scores.std()

0.020836155798953617

So by cross validation we get the correct AUC of the model, that is 0.7826 is the correct AUC with standard deviation of 0.020. That we have to use this AUC as final statistics.

In [322]:
y_freq = np.bincount(y_train)
y_val = np.nonzero(y_freq)[0]
np.vstack((y_val,y_freq[y_val])).T


array([[    0, 12197],
       [    1,  1803]], dtype=int64)

Response rate

In [323]:
1803/14000

0.12878571428571428

### Now we need to try to improve AUC by Grid search Cross validation

In [275]:
from sklearn.grid_search import GridSearchCV

In [347]:
param_dist = {"criterion": ["gini","entropy"],
              "max_depth": np.arange(3,50),
              }

tree = DecisionTreeClassifier(min_samples_split = 100,
                             min_samples_leaf = 10)

tree_cv  = GridSearchCV(tree, param_dist, cv = 10, 
                        scoring = 'roc_auc')

tree_cv.fit(X_train,y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=10, min_samples_split=100,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'criterion': ['gini', 'entropy'], 'max_depth': array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,
       37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])},
       pre_dispatch='2*n_jobs', refit=True, scoring='roc_auc', verbose=0)

Building the model using best combination of parameters

In [325]:
print("Tuned Decision Tree parameter : {}".format(tree_cv.best_params_))

Tuned Decision Tree parameter : {'criterion': 'entropy', 'max_depth': 40}


In [326]:
classifier = tree_cv.best_estimator_

classifier.fit(X_train,y_train)


DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=40,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=10, min_samples_split=100,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

#### predicting probabilities

In [327]:
y_train_prob = classifier.predict_proba(X_train)
fpr, tpr, thresholds = roc_curve(y_train, y_train_prob[:,1])
auc_d = auc(fpr, tpr)
auc_d

0.9003815891554032

In [328]:
y_test_prob = classifier.predict_proba(X_test)
fpr, tpr, thresholds = roc_curve(y_test, y_test_prob[:,1])
auc_h = auc(fpr, tpr)
auc_h

0.803691981407073

In [329]:
(auc_d-auc_h)/auc_d

0.10738736654869772

Here getting a deviation of 10.8%.  That is also more than 5%.

#### Now checking rank ordering

In [330]:
Prediction = classifier.predict_proba(X_train)
train["prob_score"] = Prediction[:,1]

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [331]:
train.head()

CUST_ID  TARGET  AGE GENDER    BALANCE OCCUPATION  SCR  HOLDING_PERIOD  \
14583   C9005       0   38      M   32662.89        SAL  486              21   
16488   C7742       0   53      M  954098.25        SAL  145              14   
11143   C3723       0   44      M   31973.61       PROF  819              10   
14011    C390       0   28      M  647270.80       SENP  811              21   
19411  C19037       1   24      M   21531.70       SENP  230               7   

      ACC_TYPE  LEN_OF_RLTN_IN_MNTH   ...    AMT_BR_CSH_WDL_DR  AMT_CHQ_DR  \
14583       SA                  149   ...                58360           0   
16488       CA                  113   ...               314470       23630   
11143       SA                  208   ...                    0      439230   
14011       SA                  204   ...                    0           0   
19411       CA                   93   ...                    0     2063920   

       AMT_NET_DR  AMT_MOB_DR   AMT_L_DR  FLG_HAS_ANY_CHGS  FLG_HAS_NOMINEE  \
14583         0.0           0    58360.0                 0                1   
16488         0.0           0   338100.0                 0                1   
11143    633187.0       20265  1094082.0                 0                0   
14011     34993.0       66584   111077.0                 0                1   
19411    242417.0      121344  2492781.0                 0                1   

       FLG_HAS_OLD_LOAN  prob_score  decile  
14583                 0    0.000000       5  
16488                 0    0.298701       9  
11143                 0    0.000000       5  
14011                 0    0.000000       5  
19411                 1    0.025000       5  

[5 rows x 27 columns]

scoring step



decile code

In [332]:
def deciles(x):
    decile = pd.Series(index=[0,1,2,3,4,5,6,7,8,9])
    for i in np.arange(0.1,1.1,0.1):
        decile[int(i*10)]=x.quantile(i)
    def z(x):
        if x<decile[1]: return(1)
        elif x<decile[2]: return(2)
        elif x<decile[3]: return(3)
        elif x<decile[4]: return(4)
        elif x<decile[5]: return(5)
        elif x<decile[6]: return(6)
        elif x<decile[7]: return(7)
        elif x<decile[8]: return(8)
        elif x<decile[9]: return(9)
        elif x<=decile[10]: return(10)
        else:return(np.NaN)
    s=x.map(z)
    return(s)

In [333]:
def Rank_Ordering(X,y,Target):
    X['decile']=deciles(X[y])
    Rank=X.groupby('decile').apply(lambda x: pd.Series([
        np.min(x[y]),
        np.max(x[y]),
        np.mean(x[y]),
        np.size(x[y]),
        np.sum(x[Target]),
        np.size(x[Target][x[Target]==0]),
        ],
        index=(["min_resp","max_resp","avg_resp",
                "cnt","cnt_resp","cnt_non_resp"])
        )).reset_index()
    Rank = Rank.sort_values(by='decile',ascending=False)
    Rank["rrate"] = Rank["cnt_resp"]*100/Rank["cnt"]
    Rank["cum_resp"] = np.cumsum(Rank["cnt_resp"])
    Rank["cum_non_resp"] = np.cumsum(Rank["cnt_non_resp"])
    Rank["cum_resp_pct"] = Rank["cum_resp"]/np.sum(Rank["cnt_resp"])
    Rank["cum_non_resp_pct"]=Rank["cum_non_resp"]/np.sum(Rank["cnt_non_resp"])
    Rank["KS"] = Rank["cum_resp_pct"] - Rank["cum_non_resp_pct"]
    Rank
    return(Rank)

In [334]:
Rank = Rank_Ordering(train,"prob_score","TARGET")
Rank

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


decile  min_resp  max_resp  avg_resp     cnt  cnt_resp  cnt_non_resp  \
5      10  0.405063  0.900000  0.551396  1469.0     810.0         659.0   
4       9  0.281690  0.400000  0.333573  1388.0     463.0         925.0   
3       8  0.160714  0.276596  0.204960  1371.0     281.0        1090.0   
2       7  0.089552  0.153846  0.114993  1374.0     158.0        1216.0   
1       6  0.027027  0.086957  0.055129  1433.0      79.0        1354.0   
0       5  0.000000  0.026316  0.001723  6965.0      12.0        6953.0   

       rrate  cum_resp  cum_non_resp  cum_resp_pct  cum_non_resp_pct        KS  
5  55.139551     810.0         659.0      0.449251          0.054030  0.395222  
4  33.357349    1273.0        1584.0      0.706045          0.129868  0.576177  
3  20.495988    1554.0        2674.0      0.861897          0.219234  0.642663  
2  11.499272    1712.0        3890.0      0.949529          0.318931  0.630598  
1   5.512910    1791.0        5244.0      0.993344          0.429942  0.563403  
0   0.172290    1803.0       12197.0      1.000000          1.000000  0.000000

We are getting 6 deciles, best Ks value is 64.26%.

Let us see the Rank Ordering on test Dataset

In [335]:
Prediction_h = classifier.predict_proba(X_test)
test["prob_score"] = Prediction_h[:,1]

Rank_h = Rank_Ordering(test,"prob_score","TARGET")
Rank_h


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


decile  min_resp  max_resp  avg_resp     cnt  cnt_resp  cnt_non_resp  \
5      10  0.416667  0.900000  0.549320   602.0     224.0         378.0   
4       9  0.285714  0.409091  0.340795   619.0     162.0         457.0   
3       8  0.179775  0.283582  0.220824   580.0     116.0         464.0   
2       7  0.094340  0.171429  0.126781   608.0      89.0         519.0   
1       6  0.040000  0.092308  0.066409   600.0      50.0         550.0   
0       5  0.000000  0.039216  0.003276  2991.0      68.0        2923.0   

       rrate  cum_resp  cum_non_resp  cum_resp_pct  cum_non_resp_pct        KS  
5  37.209302     224.0         378.0      0.315938          0.071442  0.244496  
4  26.171244     386.0         835.0      0.544429          0.157815  0.386614  
3  20.000000     502.0        1299.0      0.708039          0.245511  0.462528  
2  14.638158     591.0        1818.0      0.833568          0.343602  0.489966  
1   8.333333     641.0        2368.0      0.904090          0.447552  0.456538  
0   2.273487     709.0        5291.0      1.000000          1.000000  0.000000

Here the best KS is 48.23%, so this is good model.

In [336]:
Deviation = (0.6426 - 0.4899)/0.6426
Deviation


0.23762838468720815

It is very high, must be within 10%

In [337]:
(55.13-37.20)/55.13

0.32523127153999637

In [338]:
y_freq = np.bincount(y_test)
y_val = np.nonzero(y_freq)[0]
np.vstack((y_val,y_freq[y_val])).T

array([[   0, 5291],
       [   1,  709]], dtype=int64)

In [339]:
Responserate=709 / 6000
Responserate

0.11816666666666667

In [340]:
## LIFT For training data
0.6426/0.1287


4.9930069930069925

In [343]:
## LIFT For test data
0.4899/0.1181

4.14817950889077

In [345]:
##difference
(4.99-4.14)/4.99

0.17034068136272554

This is also very high for a model.